common/membership/ringpop/service_resolver.go

```diff
@@ -249,50 +264,41 @@ func (r *serviceResolver) HandleEvent(
 		// Note that we receive events asynchronously, possibly out of order.
 		// We cannot rely on the content of the event, rather we load everything
 		// from ringpop when we get a notification that something changed.
-		if err := r.refresh(); err != nil {
+		if err := r.refresh(refreshModeAlways); err != nil {
 			r.logger.Error("error refreshing ring when receiving a ring changed event", tag.Error(err))
 		}
 	}
 }
 
-func (r *serviceResolver) refresh() error {
+func (r *serviceResolver) refresh(mode refreshMode) error {
 	var event *membership.ChangedEvent
 	var err error
 	defer func() {
 		if event != nil {
 			r.emitEvent(event)
 		}
 	}()
-	r.refreshLock.Lock()
-	defer r.refreshLock.Unlock()
-	event, err = r.refreshLocked()
-	return err
-}
 
-func (r *serviceResolver) refreshWithBackoff() error {
-	var event *membership.ChangedEvent
-	var err error
-	defer func() {
-		if event != nil {
-			r.emitEvent(event)
-		}
-	}()
 	r.refreshLock.Lock()
 	defer r.refreshLock.Unlock()
-	if r.lastRefreshTime.After(time.Now().UTC().Add(-minRefreshInternal)) {
-		// refresh too frequently
-		return nil
+
+	if mode == refreshModeLazy && r.lastRefreshTime.After(time.Now().UTC().Add(-minRefreshInternal)) {
+		return nil // refreshed too recently
 	}
+
 	event, err = r.refreshLocked()
 	return err
 }
 
 func (r *serviceResolver) refreshLocked() (*membership.ChangedEvent, error) {
-	hosts, err := r.getReachableMembers()
+	hosts, nextEvent, err := r.getReachableMembers()
 	if err != nil {
 		return nil, err
 	}
 
+	// if we found an add/remove event, schedule another refresh right at that time
+	r.scheduleRefresh(nextEvent)
```

Comment:
As discussed offline, I think we should clamp this time to less than 10 seconds in the future, so that if there's extreme clock skew, we don't end up getting membership changes disastrously late.